## Pandas III Metodos

In [2]:
import pandas as pd

1) Cargar el dataset (Pista Si hacemos la carga del dataset sin el parámetro encoding nos puede dar error (puede que no os de error, si es así menos problemas!). Aquí tenéis una lista de los encodings disponibles en pandas. Buscad por los encoding de tipo ISO.)

In [3]:
df = pd.read_csv('attacks.csv')

2) Haced una exploración básica del DataFrame:

- ¿Cuántas filas y columnas tiene el DataFrame?

In [6]:
df.shape

(6114, 22)

Nuestro DataFrame tiene 6114 filas y 22 columnas.

- Mostrad las primeras cinco filas y las últimas cinco filas del DataFrame.

In [7]:
df.head()

Unnamed: 0 Case Number    Year        Type    Country             Area  \
0           0  2018.06.25  2018.0     Boating        USA       California   
1           1  2018.06.18  2018.0  Unprovoked        USA          Georgia   
2           2  2018.06.09  2018.0     Invalid        USA           Hawaii   
3           3  2018.06.08  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4           4  2018.06.04  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22  Unnamed: 23  
0    2018.06.25         6303.0         NaN          NaN  
1    2018.06.18         6302.0         NaN          NaN  
2    2018.06.09         6301.0         NaN          NaN  
3    2018.06.08         6300.0         NaN          NaN  
4    2018.06.04         6299.0         NaN          NaN  

[5 rows x 22 columns]

In [8]:
df.tail()

Unnamed: 0   Case Number    Year          Type    Country  \
6109        6127  1804.02.26.R  1804.0          Boat  AUSTRALIA   
6110        6128    1803.05.17  1803.0  Sea Disaster        USA   
6111        6129    1803.03.00  1803.0    Unprovoked  AUSTRALIA   
6112        6130  1802.04.13.R  1802.0    Unprovoked      INDIA   
6113        6131  1801.12.18.R  1801.0      Provoked        NaN   

                   Area                          Location  \
6109    New South Wales    Georges Head, off Port Jackson   
6110     South Carolina                    Off Charleston   
6111  Western Australia  Hamelin Harbour, at Faure Island   
6112                NaN                               NaN   
6113                NaN                               NaN   

                             Activity                             Name Sex   \
6109                              NaN                             boat  NaN   
6110                              NaN                    Captain Jones    M   
6111                              NaN  M. Lefevre & a sailor (rescuer)    M   
6112                              NaN                              NaN  NaN   
6113  Standing on landed shark's tail                 Stephen Pettigew    N   

      ...   Species                              Investigator or Source  \
6109  ...        NaN                         Sydney Gazette, 2/26/1804    
6110  ...        NaN                            Evening Post, 6/13/1803   
6111  ...        NaN  F.  Peron ref in G.P. Whitley (Fishes of Austr...   
6112  ...        NaN                  Evening Post (New York) 4/13/1802   
6113  ...  12' shark                       The Evening Post, 12/18/1801   

                                  pdf  \
6109  1804.02.26-boat-GeorgesHead.pdf   
6110             1803.05.17-Jones.pdf   
6111           1803.03.00-Lefevre.pdf   
6112           1802.04.13.R-India.pdf   
6113       1801.12.18.R-Pettigrew.pdf   

                                           href formula  \
6109  http://sharkattackfile.net/spreadsheets/pdf_di...   
6110  http://sharkattackfile.net/spreadsheets/pdf_di...   
6111  http://sharkattackfile.net/spreadsheets/pdf_di...   
6112  http://sharkattackfile.net/spreadsheets/pdf_di...   
6113  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
6109  http://sharkattackfile.net/spreadsheets/pdf_di...  1804.02.26.R   
6110  http://sharkattackfile.net/spreadsheets/pdf_di...    1803.05.17   
6111  http://sharkattackfile.net/spreadsheets/pdf_di...    1803.03.00   
6112  http://sharkattackfile.net/spreadsheets/pdf_di...  1802.04.13.R   
6113  http://sharkattackfile.net/spreadsheets/pdf_di...  1801.12.18.R   

     Case Number.2 original order Unnamed: 22  Unnamed: 23  
6109  1804.02.26.R          176.0         NaN          NaN  
6110    1803.05.17          175.0         NaN          NaN  
6111    1803.03.00          174.0         NaN          NaN  
6112  1802.04.13.R          173.0         NaN          NaN  
6113  1801.12.18.R          172.0         NaN          NaN  

[5 rows x 22 columns]

- Mostrad 2 filas al azar del DataFrame.

In [9]:
df.sample(2)

Unnamed: 0 Case Number    Year        Type      Country          Area  \
3287        3297  1973.09.00  1973.0  Unprovoked        EGYPT       Red Sea   
3832        3844  1962.01.16  1962.0    Provoked  NEW ZEALAND  South Island   

                                 Location Activity                      Name  \
3287  Southern end of the Sinai Peninsula      NaN                   Soldier   
3832               Pigeon Bay, Canterbury  Fishing  Mr. Reynish & Mr. Meikle   

     Sex   ...                                   Species   \
3287    M  ...                                Tiger shark   
3832  NaN  ...  Said to involve a 4.9 m [16'] "red shark"   

              Investigator or Source                      pdf  \
3287                          M Levy   1973.09.00-Soldier.pdf   
3832  The Sun (Australia), 1/17/1962  1962.01.16-RedShark.pdf   

                                           href formula  \
3287  http://sharkattackfile.net/spreadsheets/pdf_di...   
3832  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
3287  http://sharkattackfile.net/spreadsheets/pdf_di...    1973.09.00   
3832  http://sharkattackfile.net/spreadsheets/pdf_di...    1962.01.16   

     Case Number.2 original order Unnamed: 22  Unnamed: 23  
3287    1973.09.00         3006.0         NaN          NaN  
3832    1962.01.16         2459.0         NaN          NaN  

[2 rows x 22 columns]

- ¿Cuál es el nombre de las columnas del DataFrame?

In [52]:
columnas= df.columns
columnas

Index(['Unnamed: 0', 'Case Number', 'Year', 'Type', 'Country', 'Area',
       'Location', 'Activity', 'Name', 'Sex ', 'Age', 'Time', 'Species ',
       'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

- ¿Cuáles son los principales estádisticos del DataFrame? Mostrad por separado las variables categóricas y las variables numéricas.

In [13]:
df.describe()

Unnamed: 0         Year  original order
count  6114.000000  6114.000000     6114.000000
mean   3065.554629  1968.831207     3237.445044
std    1769.652139    43.745920     1769.652633
min       0.000000  1801.000000      172.000000
25%    1534.250000  1946.000000     1706.250000
50%    3066.500000  1980.000000     3236.500000
75%    4596.750000  2005.000000     4768.750000
max    6131.000000  2018.000000     6303.000000

In [12]:
df.describe(include= 'object')

Case Number        Type Country     Area  \
count         6113        6110    6074     5706   
unique        6113           8     201      791   
top     2018.06.25  Unprovoked     USA  Florida   
freq             1        4438    2200     1025   

                                Location Activity  Name  Sex    Age  \
count                               5626     5603  5914  5564  3448   
unique                              4020     1484  5118     6   156   
top     New Smyrna Beach, Volusia County  Surfing  male     M    17   
freq                                 162      966   511  4939   154   

             Time     Species  Investigator or Source                     pdf  \
count        2934         3409                   6098                    6114   
unique        366         1529                   4837                    6105   
top     Afternoon  White shark         C. Moore, GSAF  1898.00.00.R-Syria.pdf   
freq          183          161                     95                       2   

                                             href formula  \
count                                                6113   
unique                                               6104   
top     http://sharkattackfile.net/spreadsheets/pdf_di...   
freq                                                    2   

                                                     href Case Number.1  \
count                                                6114          6114   
unique                                               6099          6113   
top     http://sharkattackfile.net/spreadsheets/pdf_di...    1952.08.04   
freq                                                    4             2   

       Case Number.2   Unnamed: 22 Unnamed: 23  
count           6114             1           2  
unique          6114             1           2  
top       2018.06.25  stopped here      Teramo  
freq               1             1           1

- ¿Cuál es el porcentaje de valores nulos en cada columna?

In [69]:
for i in columnas:
    resultado = (df[i].isnull().sum()/len(df[i]))*100
    print(f"{i} {round(resultado,2)} %")

Unnamed: 0 0.0 %
Case Number 0.02 %
Year 0.0 %
Type 0.07 %
Country 0.65 %
Area 6.67 %
Location 7.98 %
Activity 8.36 %
Name 3.27 %
Sex  9.0 %
Age 43.6 %
Time 52.01 %
Species  44.24 %
Investigator or Source 0.26 %
pdf 0.0 %
href formula 0.02 %
href 0.0 %
Case Number.1 0.0 %
Case Number.2 0.0 %
original order 0.0 %
Unnamed: 22 99.98 %
Unnamed: 23 99.97 %


- Sacad los valores únicos y su frecuencia de las variables numéricas. Almacenad los resultados en un diccionario. Uno para cada variable.

In [75]:
df.columns

Index(['Unnamed: 0', 'Case Number', 'Year', 'Type', 'Country', 'Area',
       'Location', 'Activity', 'Name', 'Sex ', 'Age', 'Time', 'Species ',
       'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [80]:
df.dtypes

Unnamed: 0                  int64
Case Number                object
Year                      float64
Type                       object
Country                    object
Area                       object
Location                   object
Activity                   object
Name                       object
Sex                        object
Age                        object
Time                       object
Species                    object
Investigator or Source     object
pdf                        object
href formula               object
href                       object
Case Number.1              object
Case Number.2              object
original order            float64
Unnamed: 22                object
Unnamed: 23                object
dtype: object

In [9]:
valores_unicos = {col: df.value_counts() for col in df.columns if  df.dtypes[col] != 'object'}
valores_unicos


{'Unnamed: 0': Series([], dtype: int64),
 'Year': Series([], dtype: int64),
 'original order': Series([], dtype: int64)}

In [17]:
dic_unicos ={}
for i in df:
    
    if df.dtypes[i] == 'object':
        pass
    else:
         dic_unicos.update({i:(df[i].value_counts())})

In [18]:
dic_unicos

{'Unnamed: 0': 0       1
 4084    1
 4093    1
 4092    1
 4091    1
        ..
 2043    1
 2042    1
 2041    1
 2040    1
 6131    1
 Name: Unnamed: 0, Length: 6114, dtype: int64,
 'Year': 2015.0    143
 2017.0    136
 2016.0    130
 2011.0    128
 2014.0    126
          ... 
 1819.0      1
 1822.0      1
 1823.0      1
 1834.0      1
 1801.0      1
 Name: Year, Length: 207, dtype: int64,
 'original order': 569.0     2
 6303.0    1
 2208.0    1
 2210.0    1
 2211.0    1
          ..
 4260.0    1
 4261.0    1
 4262.0    1
 4263.0    1
 172.0     1
 Name: original order, Length: 6113, dtype: int64}

In [98]:
for i in df.dtypes:
    
    if i == object:
        pass
    else:
        for x in columnas:
            #print(x)
            print(df[x].value_counts())




0       1
4084    1
4093    1
4092    1
4091    1
       ..
2043    1
2042    1
2041    1
2040    1
6131    1
Name: Unnamed: 0, Length: 6114, dtype: int64
2018.06.25      1
1959.07.23.a    1
1959.05.30      1
1959.06.13      1
1959.06.14.a    1
               ..
2000.07.02.a    1
2000.07.02.b    1
2000.07.04.a    1
2000.07.04.b    1
1801.12.18.R    1
Name: Case Number, Length: 6113, dtype: int64
2015.0    143
2017.0    136
2016.0    130
2011.0    128
2014.0    126
         ... 
1819.0      1
1822.0      1
1823.0      1
1834.0      1
1801.0      1
Name: Year, Length: 207, dtype: int64
Unprovoked      4438
Provoked         564
Invalid          537
Sea Disaster     232
Boating          203
Boat             133
Questionable       2
Boatomg            1
Name: Type, dtype: int64
USA                               2200
AUSTRALIA                         1316
SOUTH AFRICA                       569
PAPUA NEW GUINEA                   130
NEW ZEALAND                        126
                     

- Cambiad el nombre de las columnas para que todas estén en minúsculas y reemplazad los espacios por _.

In [99]:
cambio_columnas = {col :  col.lower() for col in df.columns}

In [101]:
df2 = df.rename(columns = cambio_columnas) #inplace = True
df2

unnamed: 0   case number    year          type    country  \
0              0    2018.06.25  2018.0       Boating        USA   
1              1    2018.06.18  2018.0    Unprovoked        USA   
2              2    2018.06.09  2018.0       Invalid        USA   
3              3    2018.06.08  2018.0    Unprovoked  AUSTRALIA   
4              4    2018.06.04  2018.0      Provoked     MEXICO   
...          ...           ...     ...           ...        ...   
6109        6127  1804.02.26.R  1804.0          Boat  AUSTRALIA   
6110        6128    1803.05.17  1803.0  Sea Disaster        USA   
6111        6129    1803.03.00  1803.0    Unprovoked  AUSTRALIA   
6112        6130  1802.04.13.R  1802.0    Unprovoked      INDIA   
6113        6131  1801.12.18.R  1801.0      Provoked        NaN   

                   area                          location  \
0            California       Oceanside, San Diego County   
1               Georgia    St. Simon Island, Glynn County   
2                Hawaii                      Habush, Oahu   
3       New South Wales                Arrawarra Headland   
4                Colima                          La Ticla   
...                 ...                               ...   
6109    New South Wales    Georges Head, off Port Jackson   
6110     South Carolina                    Off Charleston   
6111  Western Australia  Hamelin Harbour, at Faure Island   
6112                NaN                               NaN   
6113                NaN                               NaN   

                             activity                             name sex   \
0                            Paddling                      Julie Wolfe    F   
1                            Standing                  Adyson McNeely     F   
2                             Surfing                      John Denges    M   
3                             Surfing                             male    M   
4                         Free diving                   Gustavo Ramos     M   
...                               ...                              ...  ...   
6109                              NaN                             boat  NaN   
6110                              NaN                    Captain Jones    M   
6111                              NaN  M. Lefevre & a sailor (rescuer)    M   
6112                              NaN                              NaN  NaN   
6113  Standing on landed shark's tail                 Stephen Pettigew    N   

      ...         species                              investigator or source  \
0     ...      White shark                                   R. Collier, GSAF   
1     ...              NaN                     K.McMurray, TrackingSharks.com   
2     ...              NaN                     K.McMurray, TrackingSharks.com   
3     ...        2 m shark                                     B. Myatt, GSAF   
4     ...  Tiger shark, 3m                                          A .Kipper   
...   ...              ...                                                ...   
6109  ...              NaN                         Sydney Gazette, 2/26/1804    
6110  ...              NaN                            Evening Post, 6/13/1803   
6111  ...              NaN  F.  Peron ref in G.P. Whitley (Fishes of Austr...   
6112  ...              NaN                  Evening Post (New York) 4/13/1802   
6113  ...        12' shark                       The Evening Post, 12/18/1801   

                                  pdf  \
0                2018.06.25-Wolfe.pdf   
1              2018.06.18-McNeely.pdf   
2               2018.06.09-Denges.pdf   
3            2018.06.08-Arrawarra.pdf   
4                2018.06.04-Ramos.pdf   
...                               ...   
6109  1804.02.26-boat-GeorgesHead.pdf   
6110             1803.05.17-Jones.pdf   
6111           1803.03.00-Lefevre.pdf   
6112           1802.04.13.R-India.pdf   
6113       1801.12.18.R-Pettigrew.pdf   

                                           href formu

- Eliminad las columnas de Unnamed: 22 y Unnamed: 23, Case Number.1, Case Number.2, original order, Investigator or Source, pdf, href formula.

In [102]:
df2.drop(['unnamed: 22', 'unnamed: 23', 'case number.1','case number.2', 'original order', 'investigator or source','pdf', 'href formula'], axis=1)

unnamed: 0   case number    year          type    country  \
0              0    2018.06.25  2018.0       Boating        USA   
1              1    2018.06.18  2018.0    Unprovoked        USA   
2              2    2018.06.09  2018.0       Invalid        USA   
3              3    2018.06.08  2018.0    Unprovoked  AUSTRALIA   
4              4    2018.06.04  2018.0      Provoked     MEXICO   
...          ...           ...     ...           ...        ...   
6109        6127  1804.02.26.R  1804.0          Boat  AUSTRALIA   
6110        6128    1803.05.17  1803.0  Sea Disaster        USA   
6111        6129    1803.03.00  1803.0    Unprovoked  AUSTRALIA   
6112        6130  1802.04.13.R  1802.0    Unprovoked      INDIA   
6113        6131  1801.12.18.R  1801.0      Provoked        NaN   

                   area                          location  \
0            California       Oceanside, San Diego County   
1               Georgia    St. Simon Island, Glynn County   
2                Hawaii                      Habush, Oahu   
3       New South Wales                Arrawarra Headland   
4                Colima                          La Ticla   
...                 ...                               ...   
6109    New South Wales    Georges Head, off Port Jackson   
6110     South Carolina                    Off Charleston   
6111  Western Australia  Hamelin Harbour, at Faure Island   
6112                NaN                               NaN   
6113                NaN                               NaN   

                             activity                             name sex   \
0                            Paddling                      Julie Wolfe    F   
1                            Standing                  Adyson McNeely     F   
2                             Surfing                      John Denges    M   
3                             Surfing                             male    M   
4                         Free diving                   Gustavo Ramos     M   
...                               ...                              ...  ...   
6109                              NaN                             boat  NaN   
6110                              NaN                    Captain Jones    M   
6111                              NaN  M. Lefevre & a sailor (rescuer)    M   
6112                              NaN                              NaN  NaN   
6113  Standing on landed shark's tail                 Stephen Pettigew    N   

      age           time         species   \
0      57          18h00      White shark   
1      11  14h00  -15h00              NaN   
2      48          07h45              NaN   
3     NaN            NaN        2 m shark   
4     NaN            NaN  Tiger shark, 3m   
...   ...            ...              ...   
6109  NaN            NaN              NaN   
6110  NaN            NaN              NaN   
6111  NaN            NaN              NaN   
6112  NaN            NaN              NaN   
6113  NaN            NaN        12' shark   

                                                   href  
0     http://sharkattackfile.net/spreadsheets/pdf_di...  
1     http://sharkattackfile.net/spreadsheets/pdf_di...  
2     http://sharkattackfile.net/spreadsheets/pdf_di...  
3     http://sharkattackfile.net/spreadsheets/pdf_di...  
4     http://sharkattackfile.net/spreadsheets/pdf_di...  
...                                                 ...  
6109  http://sharkattackfile.net/spreadsheets/pdf_di...  
6110  http://sharkattackfile.net/spreadsheets/pdf_di...  
6111  http://sharkattackfile.net/spreadsheets/pdf_di...  
6112  http://sharkattackfile.net/spreadsheets/pdf_di...  
6113  http://sharkattackfile.net/spreadsheets/pdf_di...  

[6114 rows x 14 columns]

- Establece la columna Case Number como índice del DataFrame.

In [103]:
df2.set_index(['case number'], inplace= True)

- Guarda el csv que no tiene esas columnas para seguir usándolo mañana.

In [105]:
open('attacks1.csv','x') 

<_io.TextIOWrapper name='attacks1.csv' mode='x' encoding='UTF-8'>

In [106]:
df2.to_csv('attacks1.csv')